testsuite/forge/src/backend/local/cargo.rs

```diff
@@ -0,0 +1,189 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::Result;
+use anyhow::{bail, Context};
+use serde::Deserialize;
+use std::{
+    env, fs,
+    path::{Path, PathBuf},
+    process::Command,
+};
+use tempfile::NamedTempFile;
+
+#[derive(Deserialize)]
+pub struct Metadata {
+    pub target_directory: PathBuf,
+    pub workspace_root: PathBuf,
+}
+
+pub fn metadata() -> Result<Metadata> {
+    let output = Command::new("cargo")
+        .arg("metadata")
+        .arg("--no-deps")
+        .arg("--format-version=1")
+        .output()
+        .context("Failed to query cargo metadata")?;
+
+    serde_json::from_slice(&output.stdout).map_err(Into::into)
+}
+
+/// Get the diem node binary from the current working directory
+pub fn get_diem_node_binary_from_worktree() -> Result<(String, PathBuf)> {
+    let metadata = metadata()?;
+    let mut revision = git_rev_parse(&metadata, "HEAD")?;
+    if git_is_worktree_dirty()? {
+        revision.push_str("-dirty");
+    }
+
+    let bin_path = cargo_build_diem_node(&metadata.workspace_root, &metadata.target_directory)?;
+
+    Ok((revision, bin_path))
+}
+
+/// This function will attempt to build the diem-node binary at an arbitrary revision.
+/// Using the `target/forge` as a working directory it will do the following:
+///     1. Look for a binary named `diem-node--<revision>`, if it already exists return it
+///     2. If the binary doesn't exist check out the revision to `target/forge/revision` by doing
+///        `git archive --format=tar <revision> | tar x`
+///     3. Using the `target/forge/target` directory as a cargo artifact directory, build the
```

Comment:
do we need to do pruning for artifact?

Reply:
Nope, since its located inside the `target` directory, all the artifacts would be wiped if someone ran a `cargo clean`

Reply:
cool, I will add it into runbook. Funny thing is before you mentioned, I did not cargo clean before 😿